### ISSUE TITLE
Allow equality comparisons with null

### ISSUE THREAD

**Issue by octogonz** [nene/eslint-plugin-no-null#1]
In our TypeScript projects, we generally use `undefined` instead of `null`.  Ideally we'd like to completely eliminate `null` everywhere, however this is impractical due to certain external libraries such as `JSON.parse()` or `react` that can return `null` from their APIs.

The `eslint-plugin-no-null` package works pretty good for this, since it ignores `null` in type annotations:

```ts
let x: string | null = JSON.parse('null');  // no error, as expected
x = null;  // lint error, as expected
doSomething(null);  // lint error, as expected
```

However, the rule also reports an error for equality tests:
```ts
x = JSON.parse('null');  // no error, as expected
if (x === null) {  // lint error, false positive <===
  . . .
}
```

In this example, the developer must suppress the lint rule, even though `null` is coming from an API that wasn't their fault.

**Proposal:**  Relax `eslint-plugin-no-null` to allow boolean comparisons.

Something equivalent to this workaround that I'm currently using:
```js
  rules: [
          {
            "selector": ":not(BinaryExpression:matches([operator='!=='], [operator='==='])) > Literal[value=null]",
            "message": "Usage of \"null\" is deprecated except when received from legacy APIs; use \"undefined\" instead"
          }
  ]
```

@nene Are you still maintaining this package? Would you accept a PR to relax the rule in this way?  If needed, we could make this behavior a configurable option.

**Comment by nene** [comment]
Hi,

That's an interesting proposal. Though I feel it's a bit of a slippery slope:

1. when you allow `null` from external API-s, they can easily sneak throughout your codebase, defeating the larger goal of getting rid of `null` inside your own code - as comparison with `null` will be fine where-ever it's encountered.
2. some external APIs will expect `null` as input. In that case the ignore-comparison-with-null won't help. IIRC, React components require returning `null` if you want to render an empty component.

In my opinion it's better to build an adapter around such external API, which would convert between null & undefined. You'd only need a bit of `// eslint-disable` comments inside the adapter, keeping the rest of the codebase null-free.

So, I'm somewhat reluctant to implement such an exception, mainly because it doesn't solve the problem 2. above. If at all, it should definitely be configurable and off by default.

**Comment by BBosman** [comment]
I would like to +1 this feature request.

I'm in the process of migrating a largish codebase from `tslint` to `eslint` and `tslint` has a `no-null-keyword`, which is similar to this rule. The difference is that it allows for strict equality checks with `null`.

The codebase itself doesn't use `null`, but when interfacing with outside parties (fetch requests/DOM access/LocalStorage/Regex/...) we need to convert from `null` to `undefined` on those edges of our system and use strict equality checks there. Having that (optional) setting would help us a lot.

**Comment by nene** [comment]
With goal of keeping this plugin plain and simple, I'm rejecting this feature request.
